### PR TITLE
Remove code paths that depend on RMM_STATIC_CUDART

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
We can remove the optimizations around  `CUDA_STATIC_RUNTIME` and instead see if the function is already in the process space so that RMM doesn't need to have any build context to run properly

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by harrism** [comment]
Are existing tests sufficient to cover this?

**Comment by robertmaynard** [comment]
> Are existing tests sufficient to cover this?

I will add explicit tests that control the type of cudart we are using to verify these chages

**Review by harrism — APPROVED** [review]

**Comment by robertmaynard** [comment]
Did more testing today and realize that more work is needed. With this PR we are getting all the symbols required at link time, but the dlsym with `RTLD_DEFAULT` is failing for a couple of the use cases. The tests still pass in CI since a `libcudart.so` exists on the machine.

Need more time to figure out what is going wrong.

**Comment by vyasr** [comment]
In what contexts is this failing? I assume it's a statically linked case, so the symbols cannot have been (dynamically) loaded with `RTLD_LOCAL` in a way that would obscure them from this scope.

**Comment by robertmaynard** [comment]
> In what contexts is this failing? I assume it's a statically linked case, so the symbols cannot have been (dynamically) loaded with `RTLD_LOCAL` in a way that would obscure them from this scope.

That is correct it is the static linking use case the is failing. The tests have the textual entries for the symbol, but the `dlsym` returns null.

File: include/rmm/detail/dynamic_load_runtime.hpp [hunk 1]
```diff
@@ -110,17 +108,30 @@ struct dynamic_load_runtime {
...
+extern "C" {
+cudaError_t cudaMemPoolCreate(cudaMemPool_t*, const cudaMemPoolProps*) __attribute((weak));
+cudaError_t cudaMemPoolSetAttribute(cudaMemPool_t, cudaMemPoolAttr, void*) __attribute((weak));
+cudaError_t cudaMemPoolDestroy(cudaMemPool_t) __attribute((weak));
+cudaError_t cudaMallocFromPoolAsync(void**, size_t, cudaMemPool_t, cudaStream_t)
+  __attribute((weak));
+cudaError_t cudaFreeAsync(void*, cudaStream_t) __attribute((weak));
+cudaError_t cudaDeviceGetDefaultMemPool_sig(cudaMemPool_t*, int) __attribute((weak));
+}
+
```
Comment (wence-) [inline include/rmm/detail/dynamic_load_runtime.hpp:121]: *question*: Why do we need to forward-declare this functions with weak symbols when previously we did not? Since this is gated behind CUDART >= 11020, I would have thought that `cuda_runtime_api.h` would be providing these forward declarations anyway?

File: include/rmm/detail/dynamic_load_runtime.hpp [hunk 2]
```diff
@@ -69,38 +69,36 @@ struct dynamic_load_runtime {
...
+    auto* error  = dlerror();
+
+    // throw rmm::logic_error{std::string{"dlysm: "} + error};
+    if (error != nullptr) {
+      // function hasn't been loaded already, load it from CUDA runtime
+      auto* runtime = get_cuda_runtime_handle();
+      handle        = ::dlsym(runtime, func_name);
+      error         = dlerror();
+    }
+    if (error != nullptr) { return std::nullopt; }
```
[the remark this answers is not in the document]
Comment (robertmaynard) [reply include/rmm/detail/dynamic_load_runtime.hpp:83]: updated to do that. They edge cases around dlsym returning a nullptr but that being valid isn't something we would hit with these functions.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -99,13 +99,23 @@ function(ConfigureTest TEST_NAME)
     set(_RMM_TEST_PERCENT 100)
   endif()
 
+  if(_RMM_TEST_CUDART STREQUAL SHARED)
+    set(cudart_link_libs $<COMPILE_ONLY:rmm> CUDA::cudart)
+  elseif(_RMM_TEST_CUDART STREQUAL STATIC)
+    set(cudart_link_libs $<COMPILE_ONLY:rmm> CUDA::cudart_static)
+  else()
+    set(cudart_link_libs rmm)
```
[the remark this answers is not in the document]
Comment (robertmaynard) [reply tests/CMakeLists.txt:107]: nvcc supports a cudart=none mode ( https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/#cudart-none-shared-static-cudart ) which is used in some embedded systems. So I opened the door for that here. Happy to drop / error if we want instead

File: include/rmm/detail/dynamic_load_runtime.hpp [hunk 4]
```diff
@@ -110,17 +108,30 @@ struct dynamic_load_runtime {
...
-#endif
+    static bool runtime_supports_pool{[] {
+      using cuda_free_async_sig      = dynamic_load_runtime::function_sig<void*, cudaStream_t>;
+      bool cuda_free_async_supported = true;
+      auto* p                        = static_cast<cuda_free_async_sig>(::cudaFreeAsync);
+      if (p == nullptr) {
+        cuda_free_async_supported =
+          dynamic_load_runtime::function<cuda_free_async_sig>("cudaFreeAsync").has_value();
+      }
+      return cuda_free_async_supported;
```
[the remark this answers is not in the document]
Comment (robertmaynard) [reply include/rmm/detail/dynamic_load_runtime.hpp]: Integrated.

File: tests/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (wence-) [reply tests/CMakeLists.txt:107]: Ah thanks, let's keep it.

File: include/rmm/detail/dynamic_load_runtime.hpp (hunk 1 above)
Comment (robertmaynard) [reply include/rmm/detail/dynamic_load_runtime.hpp:117]: I think we this approach we can remove the `CUDART >= 11020` checks.

>  Why do we need to forward-declare this functions with weak symbols when previously we did not?

Because previously we only used these functions at compile time if we built statically and against a CUDART >= 11.2. Now we are saying that when statically or shared linking against 11.2+ we don't require the symbols at runtime, so that the loader doesn't hard error when < 11.2 is used. For example you could statically link against 11.1, and fallback to runtime loading 11.8 for these symbols now ( once I remove the above ifdef guard ).

Comment (wence-) [reply include/rmm/detail/dynamic_load_runtime.hpp:117]: OK, so if I understand it, with these changes `RMM_CUDART_API_WRAPPER(some_name, some_signature)` needs to be able to resolve `some_name` at compile time (even if only as a weak symbol that will be overridden). Then, at runtime, assuming we linked against cudart dynamically rather than statically, the initial `static_cast<some_signature>(::some_name)` will return null, and we will try and dlsym then name and fail.

But, I don't understand why we need the forward declarations. If we're compiling with (shared) cudart >= 11.2, then `cuda_runtime_api.h` forward-declares the functions, so the names are defined. OK, if we compile with `--cudart none` then we get link errors because the cuda runtime header doesn't forward declare with `__attribute((weak))` but let me sweep that under the carpet.

Conversely, if we're compiling with cudart < 11.2, then the names are not defined by `cuda_runtime_api.h`, but then this code is ifdeffed away, so we don't need the foward declarations, because we never refer to the names anyway.

File: include/rmm/detail/dynamic_load_runtime.hpp [hunk 5]
```diff
@@ -110,17 +104,28 @@ struct dynamic_load_runtime {
...
  */
+
+extern "C" {
+cudaError_t cudaMemPoolCreate(cudaMemPool_t*, const cudaMemPoolProps*) __attribute((weak));
+cudaError_t cudaMemPoolSetAttribute(cudaMemPool_t, cudaMemPoolAttr, void*) __attribute((weak));
+cudaError_t cudaMemPoolDestroy(cudaMemPool_t) __attribute((weak));
+cudaError_t cudaMallocFromPoolAsync(void**, size_t, cudaMemPool_t, cudaStream_t)
+  __attribute((weak));
+cudaError_t cudaFreeAsync(void*, cudaStream_t) __attribute((weak));
+cudaError_t cudaDeviceGetDefaultMemPool_sig(cudaMemPool_t*, int) __attribute((weak));
```
Comment (wence-) [inline include/rmm/detail/dynamic_load_runtime.hpp:115]:
```suggestion
cudaError_t cudaDeviceGetDefaultMemPool(cudaMemPool_t*, int) __attribute((weak));
```

If it turns out we need to keep these, I think we need this.

Comment (wence-) [reply include/rmm/detail/dynamic_load_runtime.hpp:115]: *question*: How do we ensure that these forward declarations are the same as the symbols that the runtime actually offers? If this were linking to driver code, I would advocate using the versioned interfaces, but I don't think the runtime gives the same guarantees.

I think we are "just" lucky so far that nothing changed between cuda 11 and cuda 12 in these interfaces.

File: include/rmm/detail/dynamic_load_runtime.hpp (hunk 1 above)
Comment (robertmaynard) [reply include/rmm/detail/dynamic_load_runtime.hpp:117]: The guards around the weak forward declares and `async_alloc` need to stay in place. The `async_alloc` and related async pool allocators aren't written in a way that would allow us to remove them (functions all use types found in 11.2+)

File: include/rmm/detail/dynamic_load_runtime.hpp (hunk 5 above)
Comment (robertmaynard) [reply include/rmm/detail/dynamic_load_runtime.hpp:115]: Correct. Updated

File: include/rmm/detail/dynamic_load_runtime.hpp (hunk 1 above)
Comment (robertmaynard) [reply include/rmm/detail/dynamic_load_runtime.hpp]: > But, I don't understand why we need the forward declarations. If we're compiling with (shared) cudart >= 11.2, then cuda_runtime_api.h forward-declares the functions, so the names are defined. OK, if we compile with --cudart none then we get link errors because the cuda runtime header doesn't forward declare with __attribute((weak)) but let me sweep that under the carpet.

We can compile against 11.8 but deploy RAPIDS to a machine that has only 11.1 on it. In those cases we need to gracefully fallback at runtime instead of failing to load symbols. That is why we mark the symbols as weak, so we don't get the loader errors when the symbols aren't found with a runtime 11.1

File: include/rmm/detail/dynamic_load_runtime.hpp (hunk 5 above)
Comment (robertmaynard) [reply include/rmm/detail/dynamic_load_runtime.hpp]: > I think we are "just" lucky so far that nothing changed between cuda 11 and cuda 12 in these interfaces.

Same issue existed in the original version. The cuda runtime headers set these as macros that evaluate to the proper versioned value ( e.g. cudaMalloc_v2 or cudaFreeAsync_ptsz )

File: include/rmm/detail/dynamic_load_runtime.hpp (hunk 1 above)
Comment (wence-) [reply include/rmm/detail/dynamic_load_runtime.hpp]: Ah thanks, now I understand

**Comment by bdice** [comment]
This introduces a bit of complexity that we may be able to avoid. RMM declares its minimum supported CUDA version is 11.4. We have required a minimum of 11.2 or 11.4 everywhere I can think of for a long time. I think we can remove the shims for CUDA < 11.2 in this PR.

**Comment by robertmaynard** [comment]
> This PR has gone through a lot of churn so I haven't been keeping close track of it. @robertmaynard are there still cases that need resolving that you want help brainstorming on, or is everything resolved? Are we just waiting on reviews now?

Just waiting on reviews. Since RAPIDS has dropped the requirement to support CUDA <= 11.2 this removes almost all the need for dlsym

File: include/rmm/detail/runtime_async_alloc.hpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+   * true for `cudaMemHandleTypeNone`.
+   *
+   * @param handle_type An IPC export handle type to check for support.
+   * @return true if supported
+   * @return false if unsupported
+   */
+  static bool is_export_handle_type_supported(cudaMemAllocationHandleType handle_type)
+  {
+    int supported_handle_types_bitmask{};
+#if CUDART_VERSION >= 11030  // 11.3 introduced cudaDevAttrMemoryPoolSupportedHandleTypes
```
Comment (harrism) [inline include/rmm/detail/runtime_async_alloc.hpp:66]: Do we support CUDA < 11.3?

File: tests/mr/device/mr_ref_test.hpp [hunk 7]
```diff
@@ -166,8 +166,8 @@ inline void test_various_allocations(resource_ref ref)
   }
 }
 
-inline void test_various_async_allocations(rmm::device_async_resource_ref ref,
-                                           cuda_stream_view stream)
+inline void test_various_runtime_async_allocations(rmm::device_async_resource_ref ref,
```
Comment (harrism) [inline tests/mr/device/mr_ref_test.hpp:169]: I don't see why the function names in this header need to change. These functions are used to test all device_async resources, not just cudaMallocAsync.  So it's not clear to me what adding the word "runtime" means for these functions.

**Review by harrism — APPROVED** [review]
Should the name of this PR change now? It seems to be all about cudaMallocAsync.

I approve, but I don't really like the test function name changes that add "runtime".

File: include/rmm/mr/device/cuda_async_view_memory_resource.hpp [hunk 8]
```diff
@@ -17,9 +17,9 @@
 
 #include <rmm/cuda_device.hpp>
 #include <rmm/cuda_stream_view.hpp>
-#include <rmm/detail/dynamic_load_runtime.hpp>
 #include <rmm/detail/error.hpp>
 #include <rmm/detail/export.hpp>
+#include <rmm/detail/runtime_async_alloc.hpp>
```
Comment (vyasr) [inline include/rmm/mr/device/cuda_async_view_memory_resource.hpp:22]: Is this include being used in this file?

File: tests/prefetch_tests.cpp [hunk 9]
```diff
@@ -53,7 +53,7 @@ struct PrefetchTest : public ::testing::Test {
     // https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g8048f6ea5ad77917444567656c140c5a
     // specifically for when cudaMemRangeAttribute::cudaMemRangeAttributeLastPrefetchLocation is
     // used.
-    constexpr size_t prefetch_data_size = 4;
+    [[maybe_unused]] constexpr size_t prefetch_data_size = 4;
```
Comment (vyasr) [inline tests/prefetch_tests.cpp:56]: Could this just be inside of the if statement instead so that it will only be defined if it is used?

**Review by vyasr — APPROVED** [review]
Couple of small suggestions, otherwise LGTM.

File: include/rmm/detail/runtime_async_alloc.hpp (hunk 6 above)
Comment (robertmaynard) [reply include/rmm/detail/runtime_async_alloc.hpp:66]: https://docs.rapids.ai/install/?_gl=1*psa870*_ga*MTc0NjA1ODgyNi4xNjM5MDU3Nzg2*_ga_RKXFW6CM42*MTczMTYxMjk2Ny4yMC4wLjE3MzE2MTI5NjcuNjAuMC4w#system-req Still lists CUDA 11.2 as a supported runtime so I decided to keep this logic

File: tests/mr/device/mr_ref_test.hpp (hunk 7 above)
Comment (robertmaynard) [reply tests/mr/device/mr_ref_test.hpp:169]: I don't know why I renamed these, reverting

File: include/rmm/mr/device/cuda_async_view_memory_resource.hpp (hunk 8 above)
Comment (robertmaynard) [reply include/rmm/mr/device/cuda_async_view_memory_resource.hpp:22]: Not anymore, removed.

File: tests/prefetch_tests.cpp (hunk 9 above)
Comment (robertmaynard) [reply tests/prefetch_tests.cpp:56]: Yes it can. Updated.

File: include/rmm/detail/runtime_async_alloc.hpp (hunk 6 above)
Comment (bdice) [reply include/rmm/detail/runtime_async_alloc.hpp:66]: We should drop that. The install page is outdated.
